### PR TITLE
[mypyc] Support argument reordering in CallC

### PR DESCRIPTION
**Pull request by TH3CHARLie**
related mypyc/mypyc#734.

Support argument reordering via adding a new field to the description. It will solve the difference in args ordering between python syntax and C calling requirement(mostly with `in` ops). It should never be used together with variable args.

### CODE REVIEW

**Comment by TH3CHARLie** [comment]
for python code:
```python
def foo(x: Dict[int, int], y: int) -> bool:
    return y in x
```

before:
```c
char CPyDef_foo(PyObject *cpy_r_x, CPyTagged cpy_r_y) {
    PyObject *cpy_r_r0;
    char cpy_r_r1;
    char cpy_r_r2;
CPyL0: ;
    CPyTagged_IncRef(cpy_r_y);
    cpy_r_r0 = CPyTagged_StealAsObject(cpy_r_y);
    int __tmp1 = PyDict_Contains(cpy_r_x, cpy_r_r0);
    if (__tmp1 < 0)
        cpy_r_r1 = 2;
    else
        cpy_r_r1 = __tmp1;
    CPy_DecRef(cpy_r_r0);
    if (unlikely(cpy_r_r1 == 2)) {
        CPy_AddTraceback("../foo.py", "foo", 5, CPyStatic_globals);
        goto CPyL2;
    } else
        goto CPyL1;
CPyL1: ;
    return cpy_r_r1;
CPyL2: ;
    cpy_r_r2 = 2;
    return cpy_r_r2;
}
```

after:
```c
char CPyDef_foo(PyObject *cpy_r_x, CPyTagged cpy_r_y) {
    PyObject *cpy_r_r0;
    int32_t cpy_r_r1;
    char cpy_r_r2;
    char cpy_r_r3;
CPyL0: ;
    CPyTagged_IncRef(cpy_r_y);
    cpy_r_r0 = CPyTagged_StealAsObject(cpy_r_y);
    cpy_r_r1 = PyDict_Contains(cpy_r_x, cpy_r_r0);
    CPy_DecRef(cpy_r_r0);
    if (unlikely(cpy_r_r1 < 0)) {
        CPy_AddTraceback("../foo.py", "foo", 5, CPyStatic_globals);
        goto CPyL2;
    } else
        goto CPyL1;
CPyL1: ;
    cpy_r_r2 = cpy_r_r1;
    return cpy_r_r2;
CPyL2: ;
    cpy_r_r3 = 2;
    return cpy_r_r3;
}
```

File: mypyc/test/test_emitfunc.py [hunk 1]
```diff
@@ -235,12 +235,7 @@ def test_new_dict(self) -> None:
     def test_dict_contains(self) -> None:
         self.assert_emit_binary_op(
             'in', self.b, self.o, self.d,
-            """int __tmp1 = PyDict_Contains(cpy_r_d, cpy_r_o);
-               if (__tmp1 < 0)
-                   cpy_r_r0 = 2;
-               else
-                   cpy_r_r0 = __tmp1;
-            """)
+            """cpy_r_r0 = PyDict_Contains(cpy_r_o, cpy_r_d);""")
```
Comment (JukkaL) [inline mypyc/test/test_emitfunc.py:238]: This looks wrong. Should `assert_emit_binary_op` do the ordering as well?

File: mypyc/primitives/registry.py [hunk 2]
```diff
@@ -368,12 +370,15 @@ def c_method_op(name: str,
         truncated_type: type to truncated to(See Truncate for info)
                         if it's defined both return_type and it should be non-referenced
                         integer types or bool type
+        ordering: optional ordering of the arguments, if defined,
+                  reorders the arguments accordingly.
+                  should never be used together with var_arg_type
```
Comment (JukkaL) [inline mypyc/primitives/registry.py:375]: Also mention that all the other arguments, such as `arg_types`, are in the order accepted by the called function (i.e., after reordering).

File: mypyc/test/test_emitfunc.py (hunk 1 above)
Comment (TH3CHARLie) [reply mypyc/test/test_emitfunc.py:238]: Good catch, this indeed looks like a bug

File: mypyc/primitives/registry.py (hunk 2 above)
Comment (TH3CHARLie) [reply mypyc/primitives/registry.py:375]: I am a bit confused about this. `arg_types` are actually accepted by the python scope function(opposed to the called C function) and `ordering` is actually reordering `arg_types`

Comment (JukkaL) [reply mypyc/primitives/registry.py:375]: Sorry, I was confused. The order of all the other arguments are *before* reordering (Python ordering), right? Anyway, this illustrates that this is useful to document!

**Pull request merged**